tests/integration/security/ca_custom_root/trust_domain_validation_test.go

```diff
@@ -0,0 +1,270 @@
+				Prefix: "trust-domain-validation",
+				Inject: true,
+			})
+
+			// Deploy 3 workloads:
+			// naked: only test app without sidecar, send requests from trust domain aliases
+			// client: app with sidecar, send request from cluster.local
+			// server: app with sidecar, verify requests from cluster.local or trust domain aliases
+			var naked, client, server echo.Instance
+			echoboot.NewBuilder(ctx).
+				With(&naked, echo.Config{
+					Namespace: testNS,
+					Service:   "naked",
+					Subsets: []echo.SubsetConfig{
+						{
+							Annotations: echo.NewAnnotations().SetBool(echo.SidecarInject, false),
+						},
+					},
+				}).
+				With(&client, echo.Config{
+					Namespace: testNS,
+					Service:   "client",
+				}).
+				With(&server, echo.Config{
+					Subsets:        []echo.SubsetConfig{{}},
+					Namespace:      testNS,
+					Service:        "server",
+					ServiceAccount: true,
+					Ports: []echo.Port{
+						{
+							Name:         httpPlaintext,
+							Protocol:     protocol.HTTP,
+							ServicePort:  8090,
+							InstancePort: 8090,
+						},
+						{
+							Name:         httpMTLS,
+							Protocol:     protocol.HTTP,
+							ServicePort:  8091,
+							InstancePort: 8091,
+						},
+						{
+							Name:         tcpPlaintext,
+							Protocol:     protocol.TCP,
+							ServicePort:  8092,
+							InstancePort: 8092,
+						},
+						{
+							Name:         tcpMTLS,
+							Protocol:     protocol.TCP,
+							ServicePort:  8093,
+							InstancePort: 8093,
+						},
+					},
+					WorkloadOnlyPorts: []echo.WorkloadPort{
+						{
+							Port:     9000,
+							Protocol: protocol.TCP,
+						},
+					},
+				}).
+				BuildOrFail(t)
+
+			ctx.Config().ApplyYAMLOrFail(ctx, testNS.Name(), fmt.Sprintf(policy, testNS.Name()))
+
+			trustDomains := map[string]struct {
+				cert string
+				key  string
+			}{
+				"foo": {
+					cert: readFile(ctx, "workload-foo-cert.pem"),
+					key:  readFile(ctx, "workload-foo-key.pem"),
+				},
+				"bar": {
+					cert: readFile(ctx, "workload-bar-cert.pem"),
+					key:  readFile(ctx, "workload-bar-key.pem"),
+				},
+			}
+
+			verify := func(t *testing.T, from echo.Instance, td, port string, s scheme.Instance, allow bool) {
+				t.Helper()
+				want := "allow"
+				if !allow {
+					want = "deny"
+				}
+				name := fmt.Sprintf("%s[%s]->server:%s[%s]", from.Config().Service, td, port, want)
+				t.Run(name, func(t *testing.T) {
+					t.Helper()
+					opt := echo.CallOptions{
+						Target:   server,
+						PortName: port,
+						Host:     "server",
+						Scheme:   s,
+						Cert:     trustDomains[td].cert,
+						Key:      trustDomains[td].key,
+					}
+					retry.UntilSuccessOrFail(t, func() error {
+						// Repeat 5 times to make sure the response is stable.
+						for i := 0; i < 5; i++ {
+							var resp client2.ParsedResponses
+							var err error
+							if port == passThrough {
+								// Manually make the request for pass through port.
```

Comment:
nit/question: How come we cannot use Call here? Can we not explicitly set a port number or something?

Reply:
correct, I'm not sure if there are other limitations but the port number is preventing us to do. Currently it only supports using ports that are defined in the `Ports` list of the instance, and I just want to avoid changing too much in the test framework here.